src/components/editor/index.js

```diff
@@ -0,0 +1,261 @@
+import { PropTypes } from 'react';
+import reactStamp from 'react-stamp';
+import { fromJS, List, Repeat } from 'immutable';
+import Editor from 'draft-js-plugins-editor';
+import {
+  EditorState,
+  ContentState,
+  ContentBlock,
+  convertFromHTML,
+  convertToRaw,
+  convertFromRaw,
+  RichUtils,
+  Modifier,
+  BlockMapBuilder,
+  genKey,
+  CharacterMetadata,
+} from 'draft-js';
+
+function insertFragment(editorState, fragment, entityMap) {
+  var newContent = Modifier.replaceWithFragment(
+    editorState.getCurrentContent(),
+    editorState.getSelection(),
+    fragment
+  );
+
+  return EditorState.push(
+    editorState,
+    newContent,
+    'insert-fragment'
+  );
+}
+
+const EditorComponent = (React, ...behaviours) => reactStamp(React).compose({
+  propTypes: {
+    content: PropTypes.shape({
+      entityMap: PropTypes.object.isRequired,
+      blocks: PropTypes.array.isRequired,
+    }).isRequired,
+
+    onEditStart: PropTypes.func,
+    onEditEnd: PropTypes.func,
+    onChange: PropTypes.func,
+
+    readOnly: PropTypes.bool,
+    plugins: PropTypes.array,
+  },
+
+  init() {
+    const content = convertFromRaw(this.props.content);
+    this.state = {
+      editor: EditorState.createWithContent(content),
+      readOnly: true,
+    };
+
+    this.onClick = this.onClick.bind(this);
+    this.onChange = this.onChange.bind(this);
+    this.onFocus = this.onFocus.bind(this);
+    this.onBlur = this.onBlur.bind(this);
+
+    this.handleKeyCommand = this.handleKeyCommand.bind(this);
+    this.handleReturn = this.handleReturn.bind(this);
+    this.handlePastedText = this.handlePastedText.bind(this);
+
+    this.save = this.save.bind(this);
+  },
+
+  componentWillUnmount() {
+    this.deleteTimeout();
+    this.save();
+  },
+
+  onClick() {
+    this.setState({ readOnly: false }, () => {
+      this.refs.upstream && this.refs.upstream.focus();
```

Comment:
Perhaps I wasn't clear, because I think you misunderstood my comments.

>  The purpose of unit testing is test behavior of a particular unit, but the scope of unit testing is concerned only with public API. _focus is anything but public API, as its implementation and even the underscore-prefixed name suggest. The only public API of React components is its rendered data.

Right. That was entire reason I suggested moving the logic to an underscore-prefixed function. I suggested moving the logic to a place where we knowingly would not test it, but that if we wanted to test the *surrounding* functionality, which *is* part of the public API, we could do so by factoring it out as I outlined above: again, intentionally avoiding testing `_focus`.

> It's a bit of a mystery to me as to why you insist on having all of that unit-tested, instead of integration-tested which clearly seems to make more sense.

This entire conversation, from github to slack and back was because you said you needed a DOM to unit test this component. I said you didn't, and we've been going back and forth on that fine point ever since. That was probably myopic of us.

Setting that aside, I direct you to my actual comment: "This is arguably a case where a unit test is the wrong solution". So I am *not* insisting that everything be unit-tested. I honestly don't know where you've gotten that impression; I don't remember ever having a conversation about my test coverage requirements. In this case, I was simply responding to your comment about the refs.

---

Ok, let's take a step back. I know this has taken far longer than either of us have expected (and we'll work through that part together, offline), but I feel part of that is because we keep going in circles. Here's how I see us getting to the finish line here.

- this PR should reflect the issue spec as defined; the spec (including its implementation requirements) is not negotiable once work has started (feedback, discussion, and suggestions are encouraged and appreciated *before* work starts, though!);
- this PR should include a few [storybook stories](https://github.com/kadirahq/react-storybook) to demonstrate functionality *visually*; they are required for all React components (except when explicitly stated);
- this PR should include whatever unit tests *you feel* are required, knowing that shallow rendering is all that is presently available to you;

At that point, I can do a proper review.

Hit me up with any questions. :+1: